platforms/core-runtime/base-services/src/main/java/org/gradle/internal/service/ServiceScopeValidatorWorkarounds.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.service;
+
+import org.gradle.api.NonNullApi;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+@NonNullApi
+public class ServiceScopeValidatorWorkarounds {
```

Comment:
🤔 Why is this type named 'Workarounds`?  It should have some javadoc as a new top-level class, too.  

Or maybe it this functionality should just be moved into `ServiceScopeValidator` directly?  There doesn't seem to be much benefit to separating this, as this type and the validator itself are already so simple.

Reply:
This is a practice started at least during the work on the Configuration Cache. The idea is to clearly separate the bit that has to be addressed later on, before the feature can be considered stable.

The other reason that this is not part of the validator is because different workarounds might appear while brining the validation to other scopes and making it strict. This class then acts as a scope for those workarounds. 

This approach keeps the validator implementation clean, and very close to how it would look when this is stable, and only the `Workarounds` class will go away.